### ISSUE TITLE
Attempting to delete non-empty directory should throw `DirectoryNotEmptyException`

### ISSUE THREAD

**Issue by armanbilge** [scala-native/scala-native#2751]
It currently throws a generic `IOException`.
https://github.com/scala-native/scala-native/blob/607c81391fca7ca13f221d53a45ab60fdb03bd8b/javalib/src/main/scala/java/nio/file/Files.scala#L334-L341

**Comment by LeeTibbert** [comment]
@armanbilge Thank you for finding, reporting, & isolating this defect.

To that we get 1 worker, not 0 or 2, may I ask:  are you planning on PR'ing this or 
shall I?    


Lee

PS: you sure are knocking the bugs around!  Feels good for the 
       project to get some of the fleas knocked off.

**Comment by armanbilge** [comment]
@LeeTibbert please go ahead and PR if you wish, it would be much appreciated! :) 
Targeting the 0.4.x branch would be superb. I was not sure how to fix it without doing a lot of work.

**Comment by LeeTibbert** [comment]
I will take a look whist I am waiting for the rebase of some PRs to build.

**Comment by LeeTibbert** [comment]
OK, to be somewhat fair to SN, java 8 describes `DirectoryNotEmptyException` as
optional.  
```
    NoSuchFileException - if the file does not exist (optional specific exception)
    DirectoryNotEmptyException - if the file is a directory and could not otherwise be deleted because the directory is not empty (optional specific exception)
```
That said, I can see why both of these give information useful to the end user (or projects unit-test).

Let me explore some more.  My usual first step is WWSjD (What would Scala.js do?). 

It looks like there is code at the top of SN Files.scala which checks for the conditions
of interest: file exists, file is directory && is empty.  That code calls the delete() in
question, so I will have to trace to minimize the number of duplicate checks.  Easy
to hack code, hard to craft it.

More as it develops.

**Comment by armanbilge** [comment]
@LeeTibbert thanks! Good point about it not being required by the Java Spec :) although I think supporting this on POSIX is good form.

I was thinking the right approach would be to use `errno` to check for `ENOTEMPTY` with the `PosixException` helper.
https://github.com/scala-native/scala-native/blob/672b4a9932443df1a351decd05ea6bdf0357d1f1/javalib/src/main/scala/java/nio/file/PosixException.scala#L9-L17

**Comment by LeeTibbert** [comment]
> although I think supporting this on POSIX is good form.
I think the fix in question is javalib, which cover everything 
Files.scala covers, including POSIX.    

I need to do more tracing. Study twice, hack once ;-)   There
is some really rococo code dealing with file deletion at the 
top of javalib `Files.scala.

I am getting only odd moments this afternoon. At a first
run, I did not see a nio Files.scala for Scala.js.  I want
to poke around there some more before I charge off
and make fresh bugs.

(my first blush is that some of the rococo code at the top of Files.scala
should be moved to the `delete()` method in the same file.  Instant
intuition is good, but I like to gather at least some evidence to 
justify design moves).

**Comment by armanbilge** [comment]
> At a first run, I did not see a nio Files.scala for Scala.js

Ah, you meant that literally. I thought you meant What Would Scala.js Do in the philosophical sense :)

Note that Scala.js does not implement `java.nio` because I/O is not part of the JavaScript, although certain runtimes do offer I/O APIs. So I'm afraid you won't find any inspiration there for these sorts of things :)

However in the FS2 project we have an asynchronous files API implementation for Node.js, which does include this method.
https://github.com/typelevel/fs2/blob/6bff27df372adc948c2cc07b09d21885c1e2440a/io/js/src/main/scala/fs2/io/file/FilesPlatform.scala#L140-L141

Indeed, our implementation appropriately throws the `DirectoryNotFoundException`. Here is the relevant test, which passes on JVM and Node.js and failed on Scala Native which prompted me to open this issue.
https://github.com/typelevel/fs2/blob/6bff27df372adc948c2cc07b09d21885c1e2440a/io/shared/src/test/scala/fs2/io/file/FilesSuite.scala#L259-L265

**Comment by LeeTibbert** [comment]
Thank you for the test. That will help.

The PosixException object is a possibility to reduce code. I think an OR
on one of the cases would need to be added. remove/unlink etc
have at least one historical funny errno (two possible values for same error).

A time consuming part is figuring out what ScalaJVM reports in some
of the errno cases. I certainly hope it gives some text with the 
IOException.

My current thinking is to make a [nio] private method in File(no s)
so that the delete code is near to each other.  I think that Files(with s)
can not rely upon errno being passed under the table in the Unix case.

I think I am going to whistle past the Files.copy() code and make 
believe I never saw it.  After this, that could certainly stand some help.

So, I am still studying this.   What is your time frame, beyond day before
yesterday?

**Comment by armanbilge** [comment]
> I think that Files(with s) can not rely upon errno being passed under the table in the Unix case.

I agree with that, it's a bit too sneaky 😉 

> My current thinking is to make a [nio] private method in File(no s) so that the delete code is near to each other.

Another option I considered is to implement this in the [`UnixFileSytemProvider`](https://github.com/scala-native/scala-native/blob/63d07093f6d0a6e9de28cd8f9fb6bc1d6596c6ec/javalib/src/main/scala/scala/scalanative/nio/fs/unix/UnixFileSystemProvider.scala#L12) and delegate the `Files.delete(...)` to the `FileSystemProvider`. Perhaps then the `File` implementation can delegate to the `Files` as well.

> So, I am still studying this. What is your time frame, beyond day before yesterday?

Thanks! No rush, I believe it's better to take the time to do the right fix so we don't find ourselves revisiting this again in the future 😁 let me know how I can help.

**Comment by LeeTibbert** [comment]
How important is it that `DirectoryNotEmptyException` be thrown on Windows?

I hate 'same but slightly different quirks' implementations. Just scoping the bounds
to see what I can get away with.

```
Thanks! No rush, I believe it's better to take the time to do the right fix so we don't find ourselves revisiting this again in the future 
```
Agreed, entirely.  I think 0 revisits is a good goal, but from looking at the code I will be happy with any number with only 
two digits in it.  Some of this stuff, especially the tests looks awfully familiar.

There is an existing 'empty directory test' which does an assert on IOException.  Changing that
to `DirectoryNotEmptyException` is easy.  

I like the idea of keeping the file deletion stuff in one place `java.io.File` but a `private` routine
for use by `java.nio.file.Files` would have to be `private [java]`.  That is a pretty broad scope,
so I am looking for a better solution.

I will look at `UnixFileSystemProvider`.

**Comment by LeeTibbert** [comment]
I created a parallel 0.4.x PR. Now there are two.

I must have gotten something wrong with the 0.4.x PR because
the Linux tools test is failing with:
```
Error:  nir: Failed binary compatibility check against org.scala-native:nir_2.12:0.4.5! Found 1 potential problems
Error:   * object scala.scalanative.nir.
```
I  moved the PR to Work In Progress (WIP) until I can track down the error.
It might be something I did or it might be the previously seen & reported
binary difficulty in 0.4.5.   Either way, it takes quiet, still, focused time
to chase.

I am on the hunt for why read-only files can be opened (!!!) and need
to delay the delivery of this build failure interrupt.
